compositor_render/src/renderer/node.rs

```diff
@@ -184,17 +179,42 @@ impl Node {
     }
 }
 
-#[derive(Debug, thiserror::Error)]
-pub enum CreateNodeError {
-    #[error("Shader \"{0}\" does not exist. You have to register it first before using it in the scene definition.")]
-    ShaderNotFound(RendererId),
-
-    #[error("Error while validating the parameters for the shader node:\n{0}")]
-    ShaderNodeParametersValidationError(#[from] ParametersValidationError),
-
-    #[error("Instance of web renderer \"{0}\" does not exist. You have to register it first before using it in the scene definition.")]
-    WebRendererNotFound(RendererId),
+pub(crate) trait NodeSpecExt {
+    fn constraints<'a>(
+        &self,
+        renderers: &'a Renderers,
+    ) -> Result<&'a NodeConstraints, UpdateSceneError>;
+}
```

Comment:
For me it's a good pattern for sth that should be a method, but it can't be defined in this crate. This choice is very similar to discussion whether to use regular function or associated function.
 
- I try to use it for more helper functions.
- It's clearer on call site. Only disadvantage is that you don't know where to look for implementation, but I assume everyone is using "go to definition" in their editor, so I wouldn't expect it to be a problem.
- I wouldn't use that for any "core" logic and avoid exporting traits like that from the crate
- This could just as well be a regular funcion, type system would ensure that it's always called correctly, but this is grouping methods for that type under this type.

>  I think that those traits functions, that using it anywhere else is hardly possibly anyway, so what's the benefit here?

I think that how many places it's used it does not matter, I think it matters more here that if we add any methods to NodeSpecExt then we pollute global scope less, so it's a some form of encapsulation too.